File: 387 First Unique Character in a String.py
'''
First Unique Character in a String
https://leetcode.com/problems/first-unique-character-in-a-string/
'''
def firstUniqChar(s):
    """
    :type s: str
    :rtype: int
    """
    size = len(s)
    ref = [size] * 26
    for i,c in enumerate(s):
        d = ord(c) - ord('a')
        ref[d] = i if ref[d] == size else size + 1
    m = min(ref)
    return m if m < size else -1


s = "itwqbtcdprfsuprkrjkausiterybzncbmdvkgljxuekizvaivszowqtmrttiihervpncztuoljftlxybpgwnjb"
print firstUniqChar(s)
h = s[61]
print h
print s.index(h),s.rindex(h)

File: 387 First Unique Character in a String.java
/*
387. First Unique Character in a String
https://leetcode.com/problems/first-unique-character-in-a-string/

Given a string, find the first non-repeating character in it and return its index.
If it doesn't exist, return -1.

Examples:

s = "leetcode"
return 0.

s = "loveleetcode"
return 2.

Note: You may assume the string contains only lowercase English letters.
*/
class Solution {
    public int firstUniqChar(String s) {
        int size = s.length();
        HashMap<Character, Integer> seen = new HashMap();
        
        for (int i = 0; i < size; ++i) {
            char c = s.charAt(i);
            if (seen.containsKey(c)) {
                seen.put(c, size);
            } else{
                seen.put(c, i);
            }
        }
        
        int minIndex = size;
        for (Integer i : seen.values()) {
            minIndex = Math.min(minIndex, i);
        }
        
        return minIndex == size ? -1 : minIndex;
    }
}
